### PR TITLE
ASR component cleanups

### PR DESCRIPTION
**Pull request by juberti**
- Fix missing-token errors
- Compute latency more accurately
- Only compute WER when using file
- Shut down when audio stream ends
- Improve logging
- Tailwind tweaks

### CODE REVIEW

File: packages/ai-jsx/src/lib/asr/asr.tsx [hunk 1]
```diff
@@ -89,6 +93,13 @@ export class MicManager extends EventTarget {
     };
     const source = this.context.createMediaStreamSource(stream);
     source.connect(this.processorNode);
+    this.stream.getAudioTracks()[0].onended = () => {
+      console.log('MicManager stream ended');
+      this.stop();
+      if (onEnded) {
+        onEnded();
+      }
```
Comment (NickHeiner) [inline packages/ai-jsx/src/lib/asr/asr.tsx:101]:
```suggestion
        onEnded?.();
```

File: packages/ai-jsx/src/lib/asr/asr.tsx [hunk 2]
```diff
@@ -166,9 +177,10 @@ export type GetTokenFunction = (provider: string) => Promise<string>;
  * between when VAD detected silence and the transcript was received.
  */
 export class Transcript {
-  constructor(public text: string, public final: boolean, public latency?: number) {
+  constructor(public text: string, public final: boolean, public timestamp: number, public latency?: number) {
     this.text = text;
     this.final = final;
+    this.timestamp = timestamp;
     this.latency = latency;
```
Comment (NickHeiner) [inline packages/ai-jsx/src/lib/asr/asr.tsx:184]: I feel like none of this should be necessary. I think the `public text: string` in the constructor params is syntactic sugar for this.

File: packages/ai-jsx/src/lib/asr/asr.tsx [hunk 3]
```diff
@@ -180,7 +192,12 @@ export class Transcript {
  * speech recognition service.
  */
 export class SpeechRecognitionBase extends EventTarget {
-  private lastUtteranceEndTime: number = 0;
+  // A wall time representing when the first chunk was sent.
```
Comment (NickHeiner) [inline packages/ai-jsx/src/lib/asr/asr.tsx:195]: Use `/** */` for docstring comments so they show up in Intellisense.

File: packages/ai-jsx/src/lib/asr/asr.tsx [hunk 4]
```diff
@@ -225,47 +242,53 @@ export class SpeechRecognitionBase extends EventTarget {
...
       } else if (this.socket!.readyState == 0) {
-        this.outBuffer.push(evt.detail!);
-      } else {
-        console.error(`${this.name} socket closed`);
-      }
-    });
-    this.manager.addEventListener('vad', (evt: CustomEventInit<boolean>) => {
-      if (!evt.detail) {
-        this.lastUtteranceEndTime = performance.now();
+        // If the web socket isn't open yet, buffer the chunk.s
```
Comment (NickHeiner) [inline packages/ai-jsx/src/lib/asr/asr.tsx:259]:
```suggestion
        // If the web socket isn't open yet, buffer the chunk.
```

File: packages/ai-jsx/src/lib/asr/asr.tsx [hunk 5]
```diff
@@ -225,47 +242,53 @@ export class SpeechRecognitionBase extends EventTarget {
...
-  protected dispatchTranscript(transcript: string, final: boolean) {
-    let latency;
-    if (final) {
-      if (this.lastUtteranceEndTime == 0) {
-        console.warn(`${this.name} final transcript while VAD is still active`);
-      }
-      latency = performance.now() - this.lastUtteranceEndTime;
+  private computeLatency(recognitionMillis: number) {
+    if (!this.initialChunkMillis) {
+      return undefined;
```
Comment (NickHeiner) [inline packages/ai-jsx/src/lib/asr/asr.tsx:269]:
```suggestion
      return;
```

File: packages/nextjs-demo/src/app/asr/page.tsx [hunk 6]
```diff
@@ -36,15 +36,34 @@ interface AsrComponentProps {
...
+      return 'text-yellow-700';
+    }
+    return 'text-red-700';
+  };
+  const computeLatency = (values: number[]) => {
+    if (!values.length) {
+      return 0;
+    }
+    return values.reduce((prev, curr) => prev + curr, 0) / values.length;
+  };
```
Comment (NickHeiner) [inline packages/nextjs-demo/src/app/asr/page.tsx:62]: Looks like you just want `_.mean` here https://lodash.com/docs/4.17.15#mean

File: packages/nextjs-demo/src/app/asr/page.tsx [hunk 7]
```diff
@@ -97,35 +129,62 @@ const AsrComponent: React.FC<AsrComponentProps> = ({ name, link, id, costPerMinu
...
+  onClick,
+  disabled,
+  children,
+}) => (
+  <button
+    onClick={onClick}
+    disabled={disabled}
+    className={`ml-2 rounded-md ${
+      disabled ? 'bg-gray-300' : 'bg-fixie-fresh-salmon hover:bg-fixie-ripe-salmon'
+    } px-3 py-2 text-sm font-semibold text-white shadow-sm focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-fixie-fresh-salmon`}
```
Comment (NickHeiner) [inline packages/nextjs-demo/src/app/asr/page.tsx:169]: You can also use https://www.npmjs.com/package/classnames for this:

```ts
className={classNames('ml-2 rounded-md px-3 py-2 text-sm font-semibold text-white shadow-sm focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-fixie-fresh-salmon', disabled ? 'bg-gray-300' : 'bg-fixie-fresh-salmon hover:bg-fixie-ripe-salmon')};
```

Not hugely different but some people prefer it to the string interpolation.

**Review by NickHeiner — APPROVED** [review]

**Comment by NickHeiner** [comment]
Looks great. I was thinking it would be neat to get this into the Generic Sidekick UI (https://github.com/fixie-ai/fixie/tree/main/src/beta-frontend) so all Sidekicks get this for free.

**Review by NickHeiner — APPROVED** [review]

**Comment by juberti** [comment]
yep, that was my next step

**Pull request merged**